Microlensing and Light Curve Generation

timedilatesme opened the pull request:
This is the start of commits for the Microlensing Module Implementation that will take care of the following features:

1. Magnification Map generation using Luke's microlensing GPU code:- https://github.com/weisluke/microlensing
2. From Magnification Map to light-curve generation for various sources. The current plan is to finish it for quasars with the AGN Disk Models from Amoeba or another option is a point source model that convolves with the magnification map and produces the light curve by tracing over it.

The PR is **not open for review yet!** It will be continuously updated as new features are implemented.

@nkhadka21 can you change the name of the branch? Since our plan has changed and we are not using molet anymore! Maybe the new name for the branch could be "microlensing-integration"?

nkhadka21 commented:
Hi @timedilatesme , thank you very much for this PR! I just changed the name of the branch to microlensing. Let me know when you are ready for the PR.

timedilatesme commented:
Hello, @sibirrer  and @nkhadka21. Can you review the PR now and suggest if I am missing something? Or are there improvements that can be made to the current code?

sibirrer commented:
Hi @timedilatesme, thank you very much! The tests are failing. Have a look at the pre-commit message. Then for the CI, make sure the requirement on amoeba is given, or the import is made in a way that it is very optional (recommended) that only when the feature is requested it is made a requirement (I don't necessary want people to require to install a package if they are not making use of it)

timedilatesme commented on tests/test_param_util.py:
```diff
@@ -1,7 +1,6 @@
 import numpy as np
 import os
 from numpy import testing as npt
-from scipy.interpolate import interp1d
```
this line was removed as the pre-commit message test was failing due to this import (which isn't used in the code!)

timedilatesme commented:
Thank you so much @sibirrer, for the suggestions! I've fixed the bugs and made the imports optional, and the tests are passing now! Is it ready for review now?

sibirrer commented on slsim/Microlensing/__init__.py:
```diff
@@ -0,0 +1 @@
+LUKES_MICROLENSING_PATH = "/gpfs/home/passharma/repos/other/microlensing"
```
I don't think paths should be hard-coded. I would default to relative paths and this should be specified by a user at input

timedilatesme replied:
Now it directly imports microlensing (since it is pip installable)

sibirrer commented on slsim/lens.py:
```diff
@@ -16,6 +17,22 @@
+    # not having Luke's microlensing package installed!
+    from slsim.Microlensing.lightcurve import MicrolensingLightCurveFromLensModel
+
+    MICROLENSING_AVAILABLE = True
```
it seems none of the tests with microlensing will be executed in the CI. Can you add the microlensing code in the test_requirements.txt?

sibirrer reviewed (changes requested):
Thank you very much! I have a few comments. It also looks like the tests are not being executed (there are warnings in the code). There may be one or two functions that can not be tested without running a new map with GPUs, but I think most routines should be able to be tested by having a pre-computed micro-lensing map with a dummy test

weisluke commented on slsim/Microlensing/magmap.py:
```diff
@@ -0,0 +1,204 @@
+# Credits: Luke's Microlensing code - https://github.com/weisluke/microlensing
+from slsim.Microlensing import LUKES_MICROLENSING_PATH
+
+sys.path.append(LUKES_MICROLENSING_PATH)
```
Since I added my microlensing code to PyPI, I don't know if it's better to add that as a(n optional?) requirement, rather than appending to the system's PATH

timedilatesme replied:
done!

weisluke reviewed (commented):
Just some comments from me as I see you've been working on this :)
Apologies if I'm overstepping in any way, just wanting to provide some (hopefully useful!) input!

timedilatesme commented:
Thank you so much! @sibirrer and @weisluke for the suggestions, I will resolve most of them today!

weisluke commented on slsim/Microlensing/magmap.py:
```diff
@@ -0,0 +1,204 @@
+
+    @property
+    def pixel_size(self):
+        """Returns the pixel size in arcseconds."""
```
after a bit of testing, I think there's minimal/no worry of issues with precision actually! ignore me :)

timedilatesme commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,1103 @@
+        # Calculate 2D Gaussian map using float32
+        sigma = source_size
+        # use normal distribution for the source kernel
+        source_kernel = np.exp(-(X**2 + Y**2) / (2 * sigma**2))
```
Ahh, is it then better to just do this?:

`# make source size map
        xs = np.linspace(
            - self.magnification_map.half_length_x,
            self.magnification_map.half_length_x,
            self.magnification_map.num_pixels_x,
        )
        ys = np.linspace(
            - self.magnification_map.half_length_y,
           + self.magnification_map.half_length_y,
            self.magnification_map.num_pixels_y,
        )`

weisluke replied:
Yep! Possibly with an additional small offset depending on whether you want the value of the gaussian to be calculated at the center of the pixel or not.

timedilatesme commented on slsim/Microlensing/magmap.py:
```diff
@@ -0,0 +1,204 @@
+        :param m_solar: mass of the sun in arbitrary units. Default is 1.
+        :param m_lower: lower mass limit for the mass function in solar masses. Default is 0.08.
+        :param m_upper: upper mass limit for the mass function in solar masses. Default is 100.
+        :param light_loss: fraction of light lost due to microlensing (0 <= light_loss <= 1)
```
I have updated most of this and removed the extra ones. Is it correct now?

weisluke replied:
yep, those look fine!

nkhadka21 commented:
Hi @timedilatesme , thank you for your recent update to this PR! The current code cov for this PR is very low. Please increase the code cov. Then, I will go through the PR.

weisluke commented on slsim/Microlensing/magmap.py:
```diff
@@ -0,0 +1,192 @@
+                write_maps=False,
+                write_parities=False,
+                write_histograms=False,
+                **kwargs_IPM,
```
can leave out verbose (it's 0 by default) if you don't want to bother printing info/progress to screen
approx, write_xxx can also be left out as their defaults are fine

timedilatesme commented on slsim/Microlensing/lightcurvelensmodel.py:
```diff
@@ -0,0 +1,517 @@
+        """
+        return np.interp(time_array_new, time_array, lightcurve)
+
+    def effective_transverse_velocity_images(
```
Hi @weisluke, I have implemented the velocity function based on the papers. Can you take a look at it? Thank you very much!

timedilatesme commented on slsim/lens.py:
```diff
@@ -715,6 +765,134 @@ def extended_source_magnitude_for_each_image(self, band, lensed=False):
+                kappa_tot_images=kappa_tot_images,
+                shear_images=shear_images,
+                shear_angle_images=shear_angle_images,
+                ra_lens=self.deflector_position[0],  # TODO: check if this is correct?
```
Hi @nkhadka21 , I wanted to get the ra and dec for the deflector. Is this the correct way or should I be doing something else? also is the RA DEC in degrees or arcsec? Thank you very much!

nkhadka21 commented:
Hi @timedilatesme , it is taking very long time for the test. Not sure what is causing this.

timedilatesme commented:
Hi @nkhadka21 , Yeah I am also puzzled about this. All the microlensing tests run fine but it's stuck at test_kext_gext_distributions.py, I think this is probably something from github server side. All tests pass successfully on my local PC. I will recommit with a small update to check if it works.

timedilatesme commented on slsim/Util/astro_util.py:
```diff
@@ -391,6 +392,232 @@ def calculate_mean_time_lag(response_function):
 
 
+## The credits for the following functions go to Henry Best (https://github.com/Henry-Best-01/Amoeba)
+def pull_value_from_grid(array_2d, x_position, y_position):
```
Thanks for pointing that out! I've updated it in the code!

nkhadka21 commented:
> Hi @nkhadka21 , Yeah I am also puzzled about this. All the microlensing tests run fine but it's stuck at test_kext_gext_distributions.py, I think this is probably something from github server side. All tests pass successfully on my local PC. I will recommit with a small update to check if it works.

Yes, it might be from the github server.

timedilatesme commented:
@nkhadka21 The reason it is failing/taking a lot of time was because I recently merged slsim:main into my microlensing branch, and due to the new conventions that you defined in the latest code, some tests have issues/are taking too long.

timedilatesme commented:
Hello @sibirrer and @nkhadka21 , all tests pass now and the only place where tests are not written are the places that require GPU computation of mag maps. Can you review it now?

sibirrer commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+        """Generate lightcurves for a point source based on the convolved map.
+
+        :param source_redshift: Redshift of the source
+        :param cosmo: Cosmology object for the lens class
```
isn't that an astropy.cosmology instance?

timedilatesme replied:
Yes, I updated the description.

sibirrer commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+        :param num_lightcurves: Number of lightcurves to generate.
+            Default is 1.
+        :param x_start_position: Starting x position of the lightcurve
+            in pixel coordinates. Default is None.
```
how are pixel coordinates defined? I would rather use a unit that does not depend on numerical choices that may not be transparent to the user.

timedilatesme replied:
The pixel coordinates were supposed to be on the magnification map grid. But I agree this can be confusing, I've updated these coordinates to be in arcsec units on the magnification map grid.

sibirrer commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+            in pixel coordinates. Default is None.
+        :param y_start_position: Starting y position of the lightcurve
+            in pixel coordinates. Default is None.
+        :param phi_travel_direction: Angle of the travel direction in
```
can you define in what direction it goes and what angle=0 means?

timedilatesme replied:
Yes, since the coordinates are in the magmap frame, the angle is with respect to the x-axis of the provided magnification map. I've updated the description.

sibirrer commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+            in pixel coordinates. Default is None.
+        :param phi_travel_direction: Angle of the travel direction in
+            degrees. Default is None.
+        :param return_track_coords: Whether to return the track
```
what are 'track coordinates'?

timedilatesme replied:
These are pixel coordinates on the magnification map grid. I've updated the description.

sibirrer commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+        if not (return_track_coords) and not (return_time_array):
+            return LCs
+
+    def plot_lightcurves_and_magmap(
```
I don't think this routine needs to be inside the class. Can you put this in a separate file? Plotting and illustration serve a slightly different role.

timedilatesme replied:
I've moved it to the Plots/plot_functions.py file.

sibirrer commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+            tracks.append(np.array([x_positions, y_positions]))
+            time_arrays.append(np.linspace(0, self.time_duration, len(light_curve)))
+
+        if return_track_coords and not (return_time_array):
```
I am not a fan of having a routine having different output options. Can you return all of them and then in other routines decide what to use?

timedilatesme replied:
Done!

sibirrer commented on test_requirements.txt:
```diff
@@ -1,4 +1,5 @@
-galsim
\ No newline at end of file
+galsim
+microlensing
```
the magnification maps below are quite large (effectively larger than the entire package). Is it possible to have lower-resolution or much smaller maps for the tests?

sibirrer commented on slsim/lens.py:
```diff
@@ -732,6 +778,138 @@ def extended_source_magnitude_for_each_image(self, band, lensed=False):
+            dec_lens=dec_lens,
+            deflector_velocity_dispersion=self.deflector_velocity_dispersion(),
+            cosmology=self.cosmo,
+            kwargs_magnification_map=kwargs_microlensing["kwargs_MagnificationMap"],
```
can this be called as **kwargs_microlensing instead? I don't like reading out dictionary values as you had to make raise statements if certain quantities are not in the dictionary.

timedilatesme replied:
yes, i've updated it!

sibirrer commented on slsim/lens.py:
```diff
@@ -732,6 +778,138 @@ def extended_source_magnitude_for_each_image(self, band, lensed=False):
+            raise ValueError(
+                "kwargs_microlensing is None. Please provide a dictionary of settings required by micro-lensing calculation."
+            )
+        if "kwargs_MagnificationMap" not in kwargs_microlensing:
```
shouldn't this be better part of a lower-level definition where these parameters are asked for (or being provided as **kwargs_microlensing)

timedilatesme replied:
agreed! I've moved it!

sibirrer commented on slsim/lens.py:
```diff
@@ -732,6 +778,138 @@ def extended_source_magnitude_for_each_image(self, band, lensed=False):
+
+            kappa_star = self.kappa_star(
+                ra, dec
+            )  # TODO: why does it return a list? is it for each source?
```
is this still an issue? Should return a float if the input coordinates are floats

timedilatesme replied:
yeah it returned an array of length 1

timedilatesme replied:
But this is now redundant since I am computing everything with arrays.

sibirrer commented on slsim/lens.py:
```diff
@@ -732,6 +778,138 @@ def extended_source_magnitude_for_each_image(self, band, lensed=False):
+        shear_images = []
+        shear_angle_images = []
+
+        for i in range(len(image_positions_x)):
```
is it possible to have these routines been done as array calculations? kappa() gamma() and kappa_star() should support arrays

timedilatesme replied:
Yes, I updated the code now to have it parallely computed!

sibirrer commented on slsim/lens.py:
```diff
@@ -732,6 +778,138 @@ def extended_source_magnitude_for_each_image(self, band, lensed=False):
+        """For a given source, calculates the microlensing parameters for each
+        image position.
+
+        :param source: Source class instance
```
mismatch of input parameter definitions

timedilatesme replied:
Thanks for pointing that out! I've fixed this now!

sibirrer commented on slsim/lens.py:
```diff
@@ -732,6 +778,138 @@ def extended_source_magnitude_for_each_image(self, band, lensed=False):
+        image position.
+
+        :param source: Source class instance
+        :return: np.array(kappa_star_images),
```
can you be more explicit of what the quantities are?

timedilatesme replied:
I've updated the description!

sibirrer commented on slsim/lens.py:
```diff
@@ -696,7 +732,17 @@ def _point_source_magnitude(self, band, source_index, lensed=False, time=None):
                 )
+                if microlensing:
+                    microlensing_magnitudes = self._point_source_magnitude_microlensing(
+                        band,
```
I would explicitly assign the variables, i.e. band=band, etc to be more save

timedilatesme replied:
done!

sibirrer commented on slsim/Util/astro_util.py:
```diff
@@ -1030,3 +1032,328 @@ def convert_passband_to_nm(
+
+# Credits: Luke Weisenbach (https://github.com/weisluke/microlensing/blob/main/microlensing/Util/length_scales.py)
+def theta_star_physical(
+    z_lens: float, z_src: float, m: float = 1, cosmo=Planck18
```
do we need a default cosmology here?

timedilatesme replied:
Ahh right! We don't. I've removed it!

sibirrer commented on slsim/Util/astro_util.py:
```diff
@@ -1030,3 +1032,328 @@ def convert_passband_to_nm(
     return output_passband
+
+
+# The credits for the following function go to Henry Best (https://github.com/Henry-Best-01/Amoeba)
```
a lot of out commented text

timedilatesme replied:
Ahh true! removed it!

sibirrer commented on slsim/Microlensing/magmap.py:
```diff
@@ -0,0 +1,202 @@
+        return -2.5 * np.log10(self.magnifications / np.abs(self.mu_ave))
+
+    def get_pixel_size_meters(self, source_redshift, cosmo):
+        """Returns the pixel size in meters."""
```
input variables missing from docstring

timedilatesme replied:
fixed!

sibirrer commented on slsim/Microlensing/magmap.py:
```diff
@@ -0,0 +1,202 @@
+            cbar.set_label("Microlensing magnification")
+
+    @property
+    def mu_ave(self):
```
docstrings missing

timedilatesme replied:
fixed along with others!

sibirrer commented on slsim/Microlensing/source_morphology.py:
```diff
@@ -0,0 +1,425 @@
+__author__ = "Paras Sharma"
```
I would make a sub-folder and have one class per file

timedilatesme replied:
Done!

sibirrer commented on slsim/Microlensing/magmap.py:
```diff
@@ -0,0 +1,202 @@
+        :param kwargs_IPM: additional keyword arguments to pass to the IPM class.
+        """
+
+        self.kappa_tot = kappa_tot
```
do these variables all have to be public?

timedilatesme replied:
Not all need to be public, but the ones below are used in other classes. I've made others private!

```
self.center_x = center_x
self.center_y = center_y
self.theta_star = theta_star
self.half_length_x = half_length_x
self.half_length_y = half_length_y
self.num_pixels_x = num_pixels_x
self.num_pixels_y = num_pixels_y
self.magnifications
```

sibirrer commented on slsim/Microlensing/lightcurvelensmodel.py:
```diff
@@ -0,0 +1,519 @@
+            deflector in km/s.
+        :param cosmology: Astropy cosmology object to use for the
+            calculations.
+        :param kwargs_MagnificationMap: Keyword arguments for the
```
If you have a good pointer to the class and the class describes all the inputs well, then you do not need to have them here, as it's hard to keep the documentation up to date if things change

timedilatesme replied:
Got It! I've updated the description.

sibirrer reviewed (changes requested):
Thank you very much @timedilatesme ! I have a few requests, mostly minor, that help the readability and design of the code. It's quite a large PR and I might have missed a few things thought.

timedilatesme commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+    ):
+        """
+        :param magnification_map: MagnificationMap object, if not provided.
+        :param time_duration: Time duration for which the lightcurve is needed (in days).
```
Ahh, actually it was not (It was in source frame)! Thanks, I've fixed it now.

timedilatesme commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+            "eddington_ratio": eddington_ratio, }.
+        """
+
+        self.magnification_map = magnification_map
```
Yeah, they are not needed outside this class. I've made them private.

timedilatesme commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,384 @@
+        self,
+        return_source_morphology=False,
+    ):
+        """Get the convolved map based on the source morphology."""
```
fixed!

timedilatesme commented:
Hello @sibirrer and @nkhadka21, I've addressed all the review comments, and the tests also now have high [coverage](https://app.codecov.io/gh/LSST-strong-lensing/slsim/pull/310) (don't know why it's not showing above) except for the code blocks that produce magnification maps. Can you please take a look and let me know if any more changes are needed? If not, I think the PR is ready to be merged! (Wohoo!)

Thank you very much!

timedilatesme commented:
Hello @sibirrer and @nkhadka21, I've addressed all the review comments, and the tests also now have high coverage, except for the code blocks that produce magnification maps. Can you please take a look and let me know if any more changes are needed? If not, I think the PR is ready to be merged! (Wohoo!)

Thank you very much!

nkhadka21 commented on tests/test_Microlensing/test_lightcurve.py:
```diff
@@ -0,0 +1,366 @@
+    @pytest.mark.filterwarnings(
+        "ignore:divide by zero encountered in divide:RuntimeWarning"
+    )
+    def test_plot_lightcurves_and_magmap_runs_magnification(
```
Is it possible to move this test function to test_plot_functions.py?

timedilatesme replied:
sure!

timedilatesme replied:
I moved it but the reason I had  it in test_lightcurve.py so that I can access the predefined ml_lc instances there and now I copied these instance definitions to test_plot_functions.py

nkhadka21 reviewed (changes requested):
Hi @timedilatesme , thank you very much for your wonderful worl! I have some minor comments. Please address them!

timedilatesme commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,311 @@
+            )
+
+            # convolve the magnification map with the Gaussian kernel
+            self._convolved_map = fftconvolve(
```
cool!

timedilatesme commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,311 @@
+            rescaled_kernel_map = rescaled_kernel_map / np.nansum(rescaled_kernel_map)
+
+            # convolve the magnification map with the kernel map, #TODO: make this a cross-correlation
+            self._convolved_map = fftconvolve(
```
ok

timedilatesme commented on slsim/Microlensing/lightcurve.py:
```diff
@@ -0,0 +1,311 @@
+        source_redshift,
+        cosmo,
+        lightcurve_type="magnitude",
+        effective_transverse_velocity=1000,  # Transverse velocity in source plane (in km/s)
```
Actually in lightcurvelensmodel.py I use a velocity model and here this default value is just arbitrarily chosen. 1000 km/s is typical for relative velocity between source and deflector (plus random velocity of stars)

timedilatesme commented on slsim/Microlensing/lightcurvelensmodel.py:
```diff
@@ -0,0 +1,513 @@
+            time_array = np.array(time)
+        else:
+            raise ValueError(
+                "Time array not provided in the correct format. Please provide a time array in days."
```
Truee! Thanks for pointing that out!

timedilatesme commented on slsim/Microlensing/lightcurvelensmodel.py:
```diff
@@ -0,0 +1,513 @@
+            magnification. If 'magnification', the lightcurve is
+            returned in magnification without normalization. Default is
+            'magnitude'.
+        :param num_lightcurves: Number of lightcurves to generate.
```
done

timedilatesme commented on slsim/Microlensing/lightcurvelensmodel.py:
```diff
@@ -0,0 +1,513 @@
+        elif (isinstance(time, np.ndarray) or isinstance(time, list)) and len(
+            time
+        ) == 1:
+            lightcurve_duration = time[0]  # TODO: check if this is correct thing to do?
```
yes this is only for single time and not a lightcurve. true I can set it to 0. Thanks for pointing it out!

timedilatesme commented on slsim/Microlensing/lightcurvelensmodel.py:
```diff
@@ -0,0 +1,513 @@
+    def _interpolate_light_curve(self, lightcurve, time_array, time_array_new):
+        """Interpolate the lightcurve to a new time array.
+
+        Assuming "lightcurve" and "time_array" are 1D arrays of the same
```
done!

timedilatesme commented on slsim/Microlensing/source_morphology/agn.py:
```diff
@@ -0,0 +1,196 @@
+            # TODO: In future handle this by integrating the flux map over the band
+        # -----------------------------------------------------------
+
+        self._kernel_map = self.get_kernel_map()
```
Aah good point! It is not used in the class, I'll remove this line

timedilatesme replied:
But all other variables need to be set at init so that other functions can access them later on

timedilatesme replied:
Actually I realized it is important to define it in init so that some property functions in SourceMorphology (parent class) run properly

nkhadka21 replied:
Hi @timedilatesme , I am not sure why defining this variable here is important to SourceMorphology class because in this variable you are calling a function which is defined within this class (Am I missing anything?).

timedilatesme replied:
So in one of the tests for num_pix property below

https://github.com/timedilatesme/slsim/blob/d0723e7fa4b328f6be040660d671f43296678093/tests/test_Microlensing/test_source_morphology.py#L343-L350

I try to test the line below 

https://github.com/timedilatesme/slsim/blob/d0723e7fa4b328f6be040660d671f43296678093/slsim/Microlensing/source_morphology/source_morphology.py#L50-L51

and it turns out that `self.kernel_map` itself calls `get_kernel_map` which does not depend on num_pix for agn but it does depend on num_pix for gaussian which makes a recursion error shown below 

```
        if hasattr(gaussian_source, "_num_pix_y"):
            del gaussian_source._num_pix_y  # <-- forces np.size(self.kernel_map, 1)
    
>       _ = gaussian_source.num_pix_x  # triggers the branch setting _num_pix_x

tests/test_Microlensing/test_source_morphology.py:349: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
slsim/Microlensing/source_morphology/source_morphology.py:51: in num_pix_x
    self._num_pix_x = np.size(self.kernel_map, 0)
slsim/Microlensing/source_morphology/source_morphology.py:32: in kernel_map
    self._kernel_map = self.get_kernel_map()
slsim/Microlensing/source_morphology/gaussian.py:71: in get_kernel_map
    self.num_pix_x,
slsim/Microlensing/source_morphology/source_morphology.py:51: in num_pix_x
    self._num_pix_x = np.size(self.kernel_map, 0)
E   RecursionError: maximum recursion depth exceeded
!!! Recursion detected (same locals & position)
```


so I do need an earlier reference to self._kernel_map for this test to work

timedilatesme replied:
SO for the AGN class it is fine to remove this line but for the Gaussian Case it's a bit tricky.

timedilatesme replied:
Hi @nkhadka21, 
I did this because I wanted to use it in one of the source property functions, but I've changed that behaviour now. 

Now this line is removed! Can you merge the PR now?

nkhadka21 reviewed (commented):
Hi @timedilatesme, thank you very much for adressing all the comments! It looks fine to me. I have a very minor query. Please look into it! Then, it will be ready to merge!

nkhadka21 reviewed (approved):
Hi @timedilatesme , thank you very much for your work! It looks very good to me.

sibirrer reviewed (approved):
Looks great. Thank you very much @timedilatesme !
